### PR TITLE
[FLINK-3589] Allow setting Operator parallelism to default value

### PR DESCRIPTION
**Pull request by greghogan**
Adds the public field ExecutionConfig.PARALLELISM_UNKNOWN as a flag
value to indicate that the default parallelism should be used.

### CODE REVIEW

File: flink-core/src/main/java/org/apache/flink/api/common/ExecutionConfig.java [hunk 1]
```diff
@@ -64,10 +64,15 @@
...
 	 * The constant to use for the parallelism, if the system should use the number
-	 *  of currently available slots.
+	 * of currently available slots.
 	 */
 	public static final int PARALLELISM_AUTO_MAX = Integer.MAX_VALUE;
 
+	/**
+	 * The flag value indicating an unknown or unset parallelism.
+	 */
+	public static final int PARALLELISM_UNKNOWN = -1;
```
Comment (zentol) [inline flink-core/src/main/java/org/apache/flink/api/common/ExecutionConfig.java]: i would prefer calling it parallelism_default.

**Comment by zentol** [comment]
passing the default to Operator.setParallelism() will result in an exception, intended?

File: flink-core/src/main/java/org/apache/flink/api/common/ExecutionConfig.java (hunk 1 above)
Comment (greghogan) [inline flink-core/src/main/java/org/apache/flink/api/common/ExecutionConfig.java]: The parallelism may be overridden elsewhere and we are not resetting it back to the default. This flag simply indicates a value for which the parallelism will not be overridden.

**Comment by greghogan** [comment]
Perhaps there should be two flags, PARALLELISM_DEFAULT = -1 to reset the parallelism to the system default, and PARALLELISM_UNKNOWN = -2 which leaves the parallelism unchanged.

**Comment by zentol** [comment]
The changes look correct. If you could add a small test to verify that passing PARALLELISM_UNKNOWN doesn't modify the parallelism we're probably good to go.

**Comment by zentol** [comment]
and a rebase since github says there are conflicts ,)

**Comment by StephanEwen** [comment]
I just stumbled across this in the code.
I am a bit confused by the two different flags `PARALLELISM_DEFAULT` and `PARALLELISM_UNKNOWN`. What is the purpose of the later?

**Comment by greghogan** [comment]
`PARALLELISM_UNKNOWN` is a no-op which leaves the parallelism unchanged. This is a useful default for batch algorithms such as `JaccardIndex` for which parallelism can be optionally configured. If the user does not specify a parallelism we do not want to override the parallelism of the `ExecutionEnvironment` as would happen if we used `PARALLELISM_DEFAULT`.

**Comment by StephanEwen** [comment]
Couldn't that be equally well checked in the program that calls `setParallelism(...)`?

Just asking because this is handled only in very few parts of the DataSet API anyways (not at all handled in streaming) and seems a bit of a special case.

You may notice, I am starting to ask this question about "what needs to be part of the API, what can be part of the user program" quite a bit lately ;-) Trying to be conscious about complexity and commitment to maintaining all the API contructs.

**Comment by greghogan** [comment]
It's good to have you back :)

I think we can get by with only `PARALLELISM_DEFAULT` since `java.operators.Operator.setParallelism(int)` is only called by the Scala and Python APIs. I'll open a ticket.

**Pull request closed without merging**